slick/src/main/scala/slick/util/ManagedArrayBlockingQueue.scala

```diff
@@ -34,53 +38,50 @@ class ManagedArrayBlockingQueue[E >: Null <: PrioritizedRunnable](maximumInUse:
 
   private[this] var paused = false
 
-  private[util] def increaseInUseCount(pr: PrioritizedRunnable): Unit = {
+  private[util] def increaseInUseCount(pr: PrioritizedRunnable): Unit = locked {
     if (!pr.inUseCounterSet) {
-      locked {
-        require(inUseCount < maximumInUse, "count cannot be increased")
-        inUseCount += 1
-        pr.inUseCounterSet = true
-        if (inUseCount == maximumInUse) {
-          logger.debug("pausing")
-          paused = true
-        }
+      require(inUseCount < maximumInUse, "count cannot be increased")
+      inUseCount += 1
+      pr.inUseCounterSet = true
+      if (inUseCount == maximumInUse) {
+        logger.debug("pausing")
+        paused = true
       }
     }
   }
 
-  private[util] def decreaseInUseCount(): Unit = {
-    locked {
-      require(inUseCount > 0, "count cannot be decreased")
-      inUseCount -= 1
-      if (inUseCount == maximumInUse - 1) {
-        logger.debug("resuming")
+  private[util] def decreaseInUseCount(): Unit = locked {
+    require(inUseCount > 0, "count cannot be decreased")
+    inUseCount -= 1
+    if (inUseCount == maximumInUse - 1) {
+      logger.debug("resuming")
         paused = false
-        if (counts > 0) notEmpty.signalAll()
-      }
+      if (counts > 0) notEmpty.signalAll()
     }
   }
 
-  def offer(e: E): Boolean = {
-    checkNotNull(e)
-    checkNotInUse(e)
-    locked { insert(e) }
-  }
-
+  // implementation of offer(e), put(e) and offer(e, timeout, unit)
   private[this] def insert(e: E): Boolean = {
     val r = e.priority match {
       case WithConnection => highPrioItemQueue.insert(e)
       case Continuation => itemQueue.insert(e)
       case Fresh => if (itemQueue.count < capacity) itemQueue.insert(e) else false
     }
-    if (counts > 0) notEmpty.signal()
+    if (r) notEmpty.signal()
```

Comment:
replacing `counts > 0` with the the result of insertion looks fine, only when the queue is paused there is a possible difference between the two conditions. But since it's only signalling a condition this should be ok.